src/Symfony/Component/DependencyInjection/Compiler/ResolveHotPathPass.php

```diff
@@ -52,12 +52,12 @@ protected function processValue($value, $isRoot = false)
         if ($value instanceof ArgumentInterface) {
             return $value;
         }
-        if ($value instanceof Definition && $isRoot && (isset($this->resolvedIds[$this->currentId]) || !$value->hasTag($this->tagName))) {
-            return $value;
+        if ($value instanceof Definition && $isRoot && (isset($this->resolvedIds[$this->currentId]) || !$value->hasTag($this->tagName) || $value->isDeprecated())) {
+            return $value->isDeprecated() ? $value->clearTag($this->tagName) : $value;
         }
         if ($value instanceof Reference && ContainerBuilder::IGNORE_ON_UNINITIALIZED_REFERENCE !== $value->getInvalidBehavior() && $this->container->has($id = (string) $value)) {
             $definition = $this->container->findDefinition($id);
-            if (!$definition->hasTag($this->tagName)) {
+            if (!$definition->hasTag($this->tagName) && !$value->isDeprecated()) {
```

Comment:
```diff
- !$value->isDeprecated()
+ !$definition->isDeprecated()
```

?

But I'm still getting the deprecation with the mentioned reproducer and these changes :/

Reply:
should be good now